drv/fpga-server/src/main.rs

```diff
@@ -0,0 +1,272 @@
+            };
+            driver.configure_gpio();
+            let ecp5 = Ecp5::from(driver);
+        } else {
+            compile_error!("Board is not supported by the task/fpga");
+        }
+    }
+
+    let mut incoming = [0u8; idl::INCOMING_SIZE];
+    let mut server = ServerImpl {
+        device: ecp5,
+        device_reset_ticks: ecp5::DEVICE_RESET_DURATION,
+        application: Spi::from(SPI.get_task_id()).device(1),
+        application_reset_ticks: ecp5::APPLICATION_RESET_DURATION,
+        buffer: [0u8; 128],
+        decompressor: None,
+    };
+
+    if let Ok(DeviceState::AwaitingBitstream) = server.device.device_state() {
+        ringbuf_entry!(Trace::DeviceId(server.device.device_id().unwrap()));
+    }
+
+    loop {
+        idol_runtime::dispatch(&mut incoming, &mut server);
+    }
+}
+
+struct ServerImpl<FpgaT: Fpga> {
+    device: FpgaT,
+    device_reset_ticks: u64,
+    application: SpiDevice,
+    application_reset_ticks: u64,
+    buffer: [u8; 128],
+    decompressor: Option<gnarle::Decompressor>,
+}
+
+type RequestError = idol_runtime::RequestError<FpgaError>;
+type ReadDataLease = LenLimit<Leased<R, [u8]>, 128>;
+type WriteDataLease = LenLimit<Leased<W, [u8]>, 128>;
+
+impl<FpgaT: Fpga> idl::InOrderFpgaImpl for ServerImpl<FpgaT> {
+    fn device_enabled(&mut self, _: &RecvMessage) -> Result<u8, RequestError> {
+        Ok(self.device.device_enabled()? as u8)
+    }
+
+    fn set_device_enable(
+        &mut self,
+        _: &RecvMessage,
+        enabled: u8,
+    ) -> Result<(), RequestError> {
+        Ok(self.device.set_device_enable(enabled != 0)?)
+    }
+
+    fn reset_device(&mut self, _: &RecvMessage) -> Result<(), RequestError> {
+        Ok(self.device.reset_device(self.device_reset_ticks)?)
+    }
+
+    fn device_state(
+        &mut self,
+        _: &RecvMessage,
+    ) -> Result<DeviceState, RequestError> {
+        Ok(self.device.device_state()?)
+    }
+
+    fn device_id(&mut self, _: &RecvMessage) -> Result<u32, RequestError> {
+        Ok(self.device.device_id()?)
+    }
+
+    fn application_enabled(
+        &mut self,
+        _: &RecvMessage,
+    ) -> Result<u8, RequestError> {
+        Ok(self.device.application_enabled()? as u8)
+    }
+
+    fn set_application_enable(
+        &mut self,
+        _: &RecvMessage,
+        enabled: u8,
+    ) -> Result<(), RequestError> {
+        Ok(self.device.set_application_enable(enabled != 0)?)
+    }
+
+    fn reset_application(
+        &mut self,
+        _: &RecvMessage,
+    ) -> Result<(), RequestError> {
+        Ok(self
+            .device
+            .reset_application(self.application_reset_ticks)?)
+    }
+
+    fn start_bitstream_load(
+        &mut self,
+        _: &RecvMessage,
+        bitstream_type: BitstreamType,
+    ) -> Result<(), RequestError> {
+        if let BitstreamType::Compressed = bitstream_type {
+            self.decompressor = Some(gnarle::Decompressor::default())
```

Comment:
Should we check if `self.decompression` is `None` here, which would indicate that we're mid-load (and something is probably wrong)?

Reply:
The bitstream start/continue/finish sequence should probably be implemented as https://cliffle.com/blog/rust-typestate/ in order to force that sequence. The typestate object can then hold the decompressor and the finish step will automatically discard it.

Reply:
This has been implemented using the type state pattern and the server will return an `InvalidState` error if a client were to call these functions out of order.